### ISSUE TITLE
Why faaaaaast?

### ISSUE THREAD

**Issue by pcj** [google/containerregistry#68]
Trying to get my head around this repo.  I've copied the usage output below for these three tools.  Specifically, the `importer.par` says _Import images from a tarball into our faaaaaast format._.

Can someone explain what the "faaaaaast format" is, why it is fast, and what is that being measured relative to?

Thanks!

```sh
$  bazel run @containerregistry//:pusher.par -- --help
usage: pusher.par [-h] [--name NAME] [--tarball TARBALL] [--config CONFIG]
                  [--digest DIGEST] [--layer LAYER]
                  [--stamp-info-file STAMP_INFO_FILE] [--oci]
                  [--stderrthreshold STDERRTHRESHOLD]

Push images to a Docker Registry, faaaaaast.

optional arguments:
  -h, --help            show this help message and exit
  --name NAME           The name of the docker image to push.
  --tarball TARBALL     An optional legacy base image tarball.
  --config CONFIG       The path to the file storing the image config.
  --digest DIGEST       The list of layer digest filenames in order.
  --layer LAYER         The list of layer filenames in order.
  --stamp-info-file STAMP_INFO_FILE
                        A list of files from which to read substitutions to
                        make in the provided --name, e.g. {BUILD_USER}
  --oci                 Push the image with an OCI Manifest.
  --stderrthreshold STDERRTHRESHOLD
                        Write log events at or above this level to stderr.
```

```sh
$ bazel run @containerregistry//:puller.par -- --help
usage: puller.par [-h] [--name NAME] [--directory DIRECTORY]
                  [--stderrthreshold STDERRTHRESHOLD]

Pull images from a Docker Registry, faaaaast.

optional arguments:
  -h, --help            show this help message and exit
  --name NAME           The name of the docker image to pull and save.
                        Supports fully-qualified tag or digest references.
  --directory DIRECTORY
                        Where to save the image's files.
  --stderrthreshold STDERRTHRESHOLD
                        Write log events at or above this level to stderr.
```

```sh
$ bazel run @containerregistry//:importer.par -- --help
usage: importer.par [-h] [--tarball TARBALL] [--format {tar,tar.gz}]
                    [--directory DIRECTORY]
                    [--stderrthreshold STDERRTHRESHOLD]

Import images from a tarball into our faaaaaast format.

optional arguments:
  -h, --help            show this help message and exit
  --tarball TARBALL     The tarball containing the docker image to rewrite
                        into our fast on-disk format.
  --format {tar,tar.gz}
                        The form in which to save layers.
  --directory DIRECTORY
                        Where to save the image's files.
  --stderrthreshold STDERRTHRESHOLD
                        Write log events at or above this level to stderr.
```

**Comment by jonjohnsonjr** [comment]
See [here](https://github.com/bazelbuild/rules_docker/blob/master/intermediate-form.md).

tl;dr it avoids unnecessarily unzipping and rezipping layers, which is pretty slow in python.

**Comment by pcj** [comment]
Thanks for the reference!  I'll file other questions as I have them.
